src/lib/datocms/index.test.ts

```diff
@@ -0,0 +1,528 @@
+} from '@lib/datocms';
+
+// these imports will resolve to their mocked counterparts
+import * as env from 'astro:env/server';
+
+vi.mock('astro:env/server', async () => {
+  const actual = await import('astro:env/server');
+
+  return {
+    ...actual,
+    HEAD_START_PREVIEW: true,
+  };
+});
+
+vi.mock('../../../datocms-environment', () => ({
+  datocmsBuildTriggerId: 'mock-build-trigger-id',
+  datocmsEnvironment: 'mock-environment',
+}));
+
+const server = setupServer();
+
+beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));
+
+afterEach(() => {
+  server.resetHandlers();
+  vi.resetAllMocks();
+  vi.restoreAllMocks();
+});
+
+afterAll(() => server.close());
+
+describe('datocms:', () => {
+  describe('datocmsRequest:', () => {
+    test('should successfully fetch data', async () => {
+      const mockData = [
+        {
+          id: 1,
+          title: 'Hello World',
+          body: 'It\'s a beautiful day today.',
+        },
+      ];
+
+      server.use(
+        graphql.query('MockQuery', () => {
+          return HttpResponse.json({ data: { mockData } });
+        }),
+      );
+
+      const response = await datocmsRequest({
+        query: parse('query MockQuery { id }')
+      });
+      expect(response).toEqual({ mockData });
+    });
+
+    test('should return an empty array if no record is found', async () => {
+      const mockData: [] = [];
+
+      server.use(
+        graphql.query('MockQuery', () => {
+          return HttpResponse.json({ data: mockData });
+        }),
+      );
+
+      const response = await datocmsRequest({
+        query: parse('query MockQuery { id }')
+      });
+      expect(response).toEqual([]);
+    });
+
+    test('should throw an error if the response contains GraphQL errors', async () => {
+      const errorResponse = [{ message: 'Something went wrong' }];
+
+      server.use(
+        graphql.query('MockQuery', () => {
+          return HttpResponse.json(
+            { errors: errorResponse },
+            { status: 200 }
+          );
+        }),
+      );
+
+      let response: Error;
+
+      try {
+        await datocmsRequest({
+          query: parse('query MockQuery { id }')
+        });
+      } catch (error: unknown) {
+        response = error as Error;
+      }
+
+      expect(response!).toBeInstanceOf(Error);
+
+      // compare parsed objects instead of stringified objects because the formatting of the error message strings is an implementation detail
+      // e.g. JSON.stringify({ a: 'b' }, null, 2) does not equal JSON.stringify({ a: 'b' }, null, 4) because of the difference in whitespace
+      const parsedErrorResponse = JSON.parse(response!.message);
+      expect(parsedErrorResponse).toEqual(errorResponse);
+    });
+
+    test('should retry if it hits a rate-limit error', async () => {
+      const mockData = [
+        {
+          id: 1,
+          title: 'Hello World',
+          body: 'It\'s a beautiful day today.',
+        },
+      ];
```

Comment:
I would define this outside of the describe so it can be reused across all tests as `singleRecordCollection` for instance